Release notes 21.2.5.

stbof opened the pull request:
https://cockroachlabs.atlassian.net/browse/DOC-1892

Requires #12861, #12859, #12860 to be merged first.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/12865)
<!-- Reviewable:end -->

rmloveland commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+- Restore now inserts a system.namespace entry for synthetic public schemas. [#74759][#74759]
+- A bug has been fixed that caused internal errors in queries with set operations, like `UNION`, when corresponding columns on either side of the set operation were not the same. This error only occurred with a limited set of types. This bug is present in versions 20.2.6+, 21.1.0+, and 21.2.0+. [#75276][#75276]
+- Fixed SQL Activity pages crashing when a column was sorted by the 3rd time. [#75486][#75486]
+- Updated the `String()` function of `roleOption` to add a space on the role `VALID UNTIL`. [#75494][#75494]
```
we should link to `VALID UNTIL` in the role docs, wherever that is

rmloveland commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+- A bug has been fixed that caused internal errors in queries with set operations, like `UNION`, when corresponding columns on either side of the set operation were not the same. This error only occurred with a limited set of types. This bug is present in versions 20.2.6+, 21.1.0+, and 21.2.0+. [#75276][#75276]
+- Fixed SQL Activity pages crashing when a column was sorted by the 3rd time. [#75486][#75486]
+- Updated the `String()` function of `roleOption` to add a space on the role `VALID UNTIL`. [#75494][#75494]
+- In particular cases, some queries that involve a scan which returns many results and which includes lookups for individual keys were not returning all results from the table. [#75512][#75512]
```
should probably link 'a scan which returns many results' to https://www.cockroachlabs.com/docs/v21.2/performance-best-practices-overview.html#table-scan-best-practices

rmloveland commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+
+### Performance improvements
+
+- Allow rangefeed streams to use separate HTTP connection when `kv.rangefeed.use_dedicated_connection_class.enabled` setting is turned on.  Using separate connection class reduces the possibility of out of memory errors when running rangefeeds against very large tables.  The connection window size for rangefeed can be adjusted via `COCKROACH_RANGEFEED_INITIAL_WINDOW_SIZE` environment variable, wh ...
```
'range feed' should link to range feed docs

the cluster setting mentioned should link to cluster settings page

mention of env vars could link to https://www.cockroachlabs.com/docs/v21.2/cockroach-commands.html#environment-variables

rmloveland commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+### Performance improvements
+
+- Allow rangefeed streams to use separate HTTP connection when `kv.rangefeed.use_dedicated_connection_class.enabled` setting is turned on.  Using separate connection class reduces the possibility of out of memory errors when running rangefeeds against very large tables.  The connection window size for rangefeed can be adjusted via `COCKROACH_RANGEFEED_INITIAL_WINDOW_SIZE` environment variable, wh ...
+- Incremental `BACKUP`s now use less memory to verify coverage of prior backups. [#74588][#74588]
```
should link to backup docs

rmloveland commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+
+- Allow rangefeed streams to use separate HTTP connection when `kv.rangefeed.use_dedicated_connection_class.enabled` setting is turned on.  Using separate connection class reduces the possibility of out of memory errors when running rangefeeds against very large tables.  The connection window size for rangefeed can be adjusted via `COCKROACH_RANGEFEED_INITIAL_WINDOW_SIZE` environment variable, wh ...
+- Incremental `BACKUP`s now use less memory to verify coverage of prior backups. [#74588][#74588]
+- The merging of incremental backup layers during `RESTORE` now uses a simpler and less memory intensive algorithm. [#74593][#74593]
```
should link to `RESTORE` docs

rmloveland commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+### Contributors
+
+This release includes 64 merged PRs by 30 authors.
+We would like to thank the following contributors from the CockroachDB community:
```
"We would like to ..." through the list of CRDB team members should be deleted - I think there is a flag on the release notes generation script for this to avoid generating them in future.

rmloveland reviewed (changes requested):
Needs more ~cowbell~ links and a few minor edits

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+- Replaced raw statement with statement ID in the URL. [#75463][#75463]
+- Removed `$ internal` as one of the apps option under the Statements and Transactions page filters. [#75485][#75485]
+- Changed the order of tabs under SQL Activity page to be Statements, Transactions, and Sessions. [#75503][#75503]
+- If the user has the role `VIEWACTIVITYREDACTED`, the [statement diagnostics bundle](ui-statements-page.html#diagnostics) info on Statements page (diagnostics column), Statement Details page (diagnostics tab) and Advanced Debug page (diagnostics history) is hidden. [#75521][#75521]
```
@rmloveland Does every single mention of the Statements page in this section need a link? This seems like overkill. Isn't one link for the section sufficient?

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+- Statements are now formatted prior to being send to the UI, this is done using a new builtin function that formats statements. [#74677][#74677]
+- New role `VIEWACTIVITYREDACTED` that works similar as `VIEWACTIVITY` but restricts the usage of the [statements diagnostics bundle](ui-statements-page.html#diagnostics). It is possible for a user to have both roles (`VIEWACTIVITY` and `VIEWACTIVITYREDACTED`), but the role `VIEWACTIVITYREDACTED` takes precedent on restrictions. [#74862][#74862]
+- New role `NOSQLLOGIN` (and its inverse `SQLLOGIN`), which restricts [SQL CLI login](cockroach-sql.html#prerequisites) ability for a user while retaining their ability to log in to the DB Console (as opposed to `NOLOGIN` which restricts both SQL and DB Console). Without any role options all login behavior remains permitted as it does today. OIDC logins to the DB Console continue to be permitted  ...
+- Statements are no longer formatted prior to being sent to the UI, but the new builtin function remains. [#75463][#75463]
```
Actually both have to be removed. It was a PR that was reverted.

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+- New role `NOSQLLOGIN` (and its inverse `SQLLOGIN`), which restricts [SQL CLI login](cockroach-sql.html#prerequisites) ability for a user while retaining their ability to log in to the DB Console (as opposed to `NOLOGIN` which restricts both SQL and DB Console). Without any role options all login behavior remains permitted as it does today. OIDC logins to the DB Console continue to be permitted  ...
+- Statements are no longer formatted prior to being sent to the UI, but the new builtin function remains. [#75463][#75463]
+- Added support for SQL arrays containing JSON for in-memory processing. This does not add support for storing SQL arrays of JSON in tables. [#75529][#75529]
+- The default SQL stats flush interval is now 10 minutes. [#75524][#75524]
```
There isn't a setting.

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+- The terminate session and terminate query buttons are available to be enabled in the cluster UI. [#74530][#74530]
+- Replaced raw statement with statement ID in the URL. [#75463][#75463]
+- Removed `$ internal` as one of the apps option under the Statements and Transactions page filters. [#75485][#75485]
+- Changed the order of tabs under SQL Activity page to be Statements, Transactions, and Sessions. [#75503][#75503]
```
There isn't because it was just added to group together Statements, Transactions, and Sessions, and those existing pages were updated to say that you need to click SQL Activity first. I don't think we need a separate SQL Activity doc.

stbof replied:
Changed my mind. Added section to `ui-overview.html`.

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+
+### Enterprise edition changes
+
+- Redacted more potentially-sensitive URI elements from changefeed job descriptions. This is a breaking change for workflows involving copying URIs. As an alternative, the unredacted URI may be accessed from the jobs table directly. [#75187][#75187]
```
Fixed.

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+
+### SQL language changes
+
+- Statements are now formatted prior to being send to the UI, this is done using a new builtin function that formats statements. [#74677][#74677]
```
Fixed.

stbof commented on releases/v21.2.5.md:
```diff
@@ -0,0 +1,150 @@
+### SQL language changes
+
+- Statements are now formatted prior to being send to the UI, this is done using a new builtin function that formats statements. [#74677][#74677]
+- New role `VIEWACTIVITYREDACTED` that works similar as `VIEWACTIVITY` but restricts the usage of the [statements diagnostics bundle](ui-statements-page.html#diagnostics). It is possible for a user to have both roles (`VIEWACTIVITY` and `VIEWACTIVITYREDACTED`), but the role `VIEWACTIVITYREDACTED` takes precedent on restrictions. [#74862][#74862]
```
Fixed.

stbof commented:
@rmloveland Thank you for the very thorough review. Hopefully I'll do better in my first draft of the next release notes! I believe I've addressed all your comments.

rmloveland reviewed (approved):
LGTM thanks for the updates @stbof 

Ya we def do add a lot of links to these - hopefully we can automate some of this in future, it's pretty tedious - I think @nickvigilante had some ideas when we did a release notes edit together recently

nickvigilante commented:
@rmloveland I'm confident we will. I'll need help in some parts, but it'll get done.

@stbof For now, don't forget to update the sidebar-releases.json file and the release index.

stbof commented:
> @rmloveland I'm confident we will. I'll need help in some parts, but it'll get done.
> 
> @stbof For now, don't forget to update the sidebar-releases.json file and the release index.

Thanks for that pointer. `sidebar-releases.json` is not mentioned here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/186548364/Release+notes

stbof commented:
Replaced with #12901.
